bin/varnishd/cache/cache_vrt.c

```diff
@@ -989,7 +989,8 @@ VRT_VSA_GetPtr(VRT_CTX, const struct suckaddr *sua, const unsigned char ** dst)
 }
 
 void
-VRT_Format_Proxy(struct vsb *vsb, VCL_INT version, VCL_IP sac, VCL_IP sas)
+VRT_Format_Proxy(struct vsb *vsb, VCL_INT version, VCL_IP sac, VCL_IP sas,
+    VCL_STRING auth)
```

Comment:
Should we make the last argument generic?

```c
VRT_Format_Proxy(vsb, version, sac, sas, "auth=varnish.test[,other=value,...]");
```

This way we future-proof the addition of future TLVs without planning to break VRT.

Reply:
Is a parsed sting the best option? Because the TLVs are well defined, I would have thought passing a struct would be appropriate?

Reply:
Well, on the VCL side (via `vmod_proxy`) I'd rather avoid pushing the knowledge of what TLVs we support. But something like this could work too:

```
$Function BLOB header(INT version, IP client, IP server, [STRING authority])
```

And in the future we could add other `[optional]` arguments if we want to support more TLVs.

Reply:
see https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt#L535
The TLVs do not need to be strings, though I agree that the ones which I would consider useful at this point in time are: `PP2_TYPE_ALPN`, `PP2_TYPE_AUTHORITY` and `PP2_TYPE_NETNS`
But already if we wanted to generate `PP2_TYPE_SSL`, a string serialization would not seem to make much sense, IMHO.
I think, for a VRT function to be useful, it should actually save some work for the caller, otherwise it would make more sense for a caller to fill out the tlv structures and just pass them.

Reply:
TLVs don't need to be strings, but they can be represented as such. Maybe this is a good case for VMOD varargs?

```
$Function BLOB header(INT version, IP client, IP server, STRING... tlv)
```

Where the TLV vararg expects a TLV name followed by the value:

```vcl
proxy.header(1, client.ip, server.ip);
proxy.header(2, client.ip, server.ip, "authority", "example.com", "alpn", "http/1");
proxy.header(2, client.ip, server.ip, "ssl", "client,cert", "ssl.version", "TLS1.2"); # involves some parsing
```

Before you say anything, I know, evil.

By the way, do we want `ENUM { 1, 2 } version`? `vmodtool.py` will happilly let you but the libvcc parser will expect an ID token. `ENUM { v1, v2 }`?

Reply:
@Dridi please let us not create a dependency on yet another feature which might take even more time to get in, after this ticket will celebrate its first anniversary tomorrow.
My suggestion at this point is to mark `VRT_Format_Proxy()` as having a transitional interface and come up with more improvements later. Right now, there is exactly one use case and I do not understand why we would need to keep back that one use case just because we do not know how the ideal interface would look like for hypothetical use cases.
Regarding representing stuff as strings: Sure, we all know what serialization is, but really do we want to `blob.encode(BASE64)` arguments to `proxy.header`?
The enum idea is a good one, thank you!

Reply:
@Dridi I feel an urge to pick on one aspect: The suggestion to add this to vmod_proxy comes from here: https://github.com/varnishcache/varnish-cache/pull/2845#issuecomment-441276492 If the interface question turned out to be holding this up, I would rather move the vmod function to vmod_debug again in order to get ahead. For now, it really only exists for coverage testing.